File: scripts/fastlio_to_mavros_only_pose.py
#!/usr/bin/env python3
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, HistoryPolicy, DurabilityPolicy
from nav_msgs.msg import Odometry
from geometry_msgs.msg import PoseWithCovarianceStamped

class FastLioVisionBridge(Node):
    def __init__(self):
        super().__init__('fastlio_vision_bridge')

        # 宣告參數
        self.declare_parameter('fastlio_odom_topic', '/Odometry')
        self.declare_parameter('mavros_vision_pose_topic', '/mavros/vision_pose/pose_cov')
        self.declare_parameter('output_frame_id', 'odom')
        
        # 協方差縮放參數
        self.declare_parameter('position_cov_scale', 1000.0)      # 位置協方差放大倍數
        self.declare_parameter('orientation_cov_scale', 1000.0)   # 姿態協方差放大倍數
        self.declare_parameter('min_position_cov', 0.01)          # 最小位置協方差
        self.declare_parameter('min_orientation_cov', 0.01)       # 最小姿態協方差

        # 讀取參數
        fastlio_topic = self.get_parameter('fastlio_odom_topic').value
        mavros_topic = self.get_parameter('mavros_vision_pose_topic').value
        self.output_frame = self.get_parameter('output_frame_id').value

        self.pos_scale = self.get_parameter('position_cov_scale').value
        self.ori_scale = self.get_parameter('orientation_cov_scale').value
        self.min_pos_cov = self.get_parameter('min_position_cov').value
        self.min_ori_cov = self.get_parameter('min_orientation_cov').value
        
        # QoS 設定
        qos_profile = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )

        # 訂閱 ZED Odometry
        self.sub = self.create_subscription(
            Odometry,
            fastlio_topic,
            self.odom_callback,
            10
        )

        # 發布到 MAVROS
        self.pub = self.create_publisher(
            PoseWithCovarianceStamped,
            mavros_topic,
            qos_profile
        )
        
        # 初始化計數器
        self.msg_count = 0

        # 輸出啟動資訊
        self.get_logger().info('=' * 50)
        self.get_logger().info('FAST-LIO2 Vision Bridge Started (Pose Only)')
        self.get_logger().info(f'  Subscribe: {fastlio_topic}')
        self.get_logger().info(f'  Publish:   {mavros_topic}')
        self.get_logger().info(f'  Frame ID:  {self.output_frame}')
        self.get_logger().info(f'  Position Cov Scale: {self.pos_scale}')
        self.get_logger().info(f'  Orientation Cov Scale: {self.ori_scale}')
        self.get_logger().info(f'  Min Position Cov: {self.min_pos_cov}')
        self.get_logger().info(f'  Min Orientation Cov: {self.min_ori_cov}')
        self.get_logger().info('=' * 50)

    def scale_covariance(self, original_cov):

        scaled_cov = list(original_cov)
        
        # 位置協方差 (對角線 0, 7, 14)
        for i in [0, 7, 14]:
            scaled_cov[i] = max(original_cov[i] * self.pos_scale, self.min_pos_cov)
        
        # 姿態協方差 (對角線 21, 28, 35)
        for i in [21, 28, 35]:
            scaled_cov[i] = max(original_cov[i] * self.ori_scale, self.min_ori_cov)
        
        # 非對角線元素也要適當縮放（保持相關性）
        for i in range(36):
            if i not in [0, 7, 14, 21, 28, 35]:
                # 根據行列位置決定用哪個縮放因子
                row = i // 6
                col = i % 6
                if row < 3 and col < 3:
                    scaled_cov[i] = original_cov[i] * self.pos_scale
                elif row >= 3 and col >= 3:
                    scaled_cov[i] = original_cov[i] * self.ori_scale
                else:
                    # 位置-姿態交叉項，用幾何平均
                    scaled_cov[i] = original_cov[i] * ((self.pos_scale * self.ori_scale) ** 0.5)
        
        return scaled_cov
        
    def odom_callback(self, msg: Odometry):
        # 建立輸出訊息
        vision_msg = PoseWithCovarianceStamped()

        # 保留 FAST-LIO2 原始時間戳（支援 EKF 延遲補償）
        vision_msg.header.stamp = msg.header.stamp
        
        # 覆蓋 Frame ID
        vision_msg.header.frame_id = self.output_frame

        # 複製位姿
        vision_msg.pose.pose = msg.pose.pose
        
        # 縮放協方差
        vision_msg.pose.covariance = self.scale_covariance(msg.pose.covariance)

        # 發布訊息
        self.pub.publish(vision_msg)

        # 定期輸出狀態
        self.msg_count += 1
        if self.msg_count % 10 == 0:
            pos = msg.pose.pose.position
            stamp = msg.header.stamp
            
            orig_cov = msg.pose.covariance[0]
            scaled_cov = vision_msg.pose.covariance[0]
            
            self.get_logger().info(
                f'[{self.msg_count}] '
                f'Time: {stamp.sec}.{stamp.nanosec // 1000000:03d} | '
                f'Pos: ({pos.x:.3f}, {pos.y:.3f}, {pos.z:.3f}) | '
                f'Cov: {orig_cov:.6f} -> {scaled_cov:.4f}'
            )


def main(args=None):
    rclpy.init(args=args)
    node = FastLioVisionBridge()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        node.get_logger().info('Received shutdown signal...')
    finally:
        node.get_logger().info('Bridge stopped.')
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/fastlio_to_mavros_only_pose.cpp
#include <rclcpp/rclcpp.hpp>
#include <rclcpp/qos.hpp>
#include <nav_msgs/msg/odometry.hpp>
#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
#include <array>
#include <cmath>

class FastLioVisionBridge : public rclcpp::Node
{
public:
  FastLioVisionBridge() : Node("fastlio_vision_bridge"), msg_count_(0)
  {
    declare_parameter("fastlio_odom_topic", "/Odometry");
    declare_parameter("mavros_vision_pose_topic", "/mavros/vision_pose/pose_cov");
    declare_parameter("output_frame_id", "odom");
    declare_parameter("position_cov_scale", 1000.0);
    declare_parameter("orientation_cov_scale", 1000.0);
    declare_parameter("min_position_cov", 0.01);
    declare_parameter("min_orientation_cov", 0.01);

    const auto fastlio_topic  = get_parameter("fastlio_odom_topic").as_string();
    const auto mavros_topic   = get_parameter("mavros_vision_pose_topic").as_string();
    output_frame_   = get_parameter("output_frame_id").as_string();
    pos_scale_      = get_parameter("position_cov_scale").as_double();
    ori_scale_      = get_parameter("orientation_cov_scale").as_double();
    min_pos_cov_    = get_parameter("min_position_cov").as_double();
    min_ori_cov_    = get_parameter("min_orientation_cov").as_double();

    sub_ = create_subscription<nav_msgs::msg::Odometry>(
      fastlio_topic, 10,
      std::bind(&FastLioVisionBridge::odomCallback, this, std::placeholders::_1));

    rclcpp::QoS qos(rclcpp::KeepLast(10));
    qos.reliability(rclcpp::ReliabilityPolicy::Reliable);
    qos.durability(rclcpp::DurabilityPolicy::Volatile);

    pub_ = create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(mavros_topic, qos);

    RCLCPP_INFO(get_logger(), "==================================================");
    RCLCPP_INFO(get_logger(), "FAST-LIO2 Vision Bridge Started (Pose Only)");
    RCLCPP_INFO(get_logger(), "  Subscribe: %s", fastlio_topic.c_str());
    RCLCPP_INFO(get_logger(), "  Publish:   %s", mavros_topic.c_str());
    RCLCPP_INFO(get_logger(), "  Frame ID:  %s", output_frame_.c_str());
    RCLCPP_INFO(get_logger(), "  Position Cov Scale: %.1f", pos_scale_);
    RCLCPP_INFO(get_logger(), "  Orientation Cov Scale: %.1f", ori_scale_);
    RCLCPP_INFO(get_logger(), "  Min Position Cov: %.4f", min_pos_cov_);
    RCLCPP_INFO(get_logger(), "  Min Orientation Cov: %.4f", min_ori_cov_);
    RCLCPP_INFO(get_logger(), "==================================================");
  }

private:
  using CovArray = std::array<double, 36>;

  CovArray scaleCovariance(const CovArray & orig) const
  {
    CovArray scaled = orig;

    // diagonal: position (0,7,14), orientation (21,28,35)
    for (int i : {0, 7, 14}) {
      scaled[i] = std::max(orig[i] * pos_scale_, min_pos_cov_);
    }
    for (int i : {21, 28, 35}) {
      scaled[i] = std::max(orig[i] * ori_scale_, min_ori_cov_);
    }

    const double cross_scale = std::sqrt(pos_scale_ * ori_scale_);

    for (int i = 0; i < 36; ++i) {
      if (i == 0 || i == 7 || i == 14 || i == 21 || i == 28 || i == 35) continue;
      const int row = i / 6;
      const int col = i % 6;
      if (row < 3 && col < 3) {
        scaled[i] = orig[i] * pos_scale_;
      } else if (row >= 3 && col >= 3) {
        scaled[i] = orig[i] * ori_scale_;
      } else {
        scaled[i] = orig[i] * cross_scale;
      }
    }

    return scaled;
  }

  void odomCallback(const nav_msgs::msg::Odometry::SharedPtr msg)
  {
    geometry_msgs::msg::PoseWithCovarianceStamped vision_msg;
    vision_msg.header.stamp    = msg->header.stamp;
    vision_msg.header.frame_id = output_frame_;
    vision_msg.pose.pose       = msg->pose.pose;
    vision_msg.pose.covariance = scaleCovariance(msg->pose.covariance);

    pub_->publish(vision_msg);

    if (++msg_count_ % 10 == 0) {
      const auto & pos   = msg->pose.pose.position;
      const auto & stamp = msg->header.stamp;
      RCLCPP_INFO(
        get_logger(),
        "[%d] Time: %d.%03d | Pos: (%.3f, %.3f, %.3f) | Cov: %.6f -> %.4f",
        msg_count_,
        stamp.sec, stamp.nanosec / 1000000,
        pos.x, pos.y, pos.z,
        msg->pose.covariance[0],
        vision_msg.pose.covariance[0]);
    }
  }

  rclcpp::Subscription<nav_msgs::msg::Odometry>::SharedPtr sub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr pub_;

  std::string output_frame_;
  double pos_scale_;
  double ori_scale_;
  double min_pos_cov_;
  double min_ori_cov_;
  int msg_count_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<FastLioVisionBridge>();
  try {
    rclcpp::spin(node);
  } catch (const std::exception & e) {
    RCLCPP_ERROR(node->get_logger(), "Exception: %s", e.what());
  }
  RCLCPP_INFO(node->get_logger(), "Bridge stopped.");
  rclcpp::shutdown();
  return 0;
}
